rectools/models/sasrec.py

```diff
@@ -0,0 +1,723 @@
+        """
+        Pass user history through item embeddings and transformer blocks.
+
+        Returns
+        -------
+            torch.Tensor. [batch_size, session_maxlen, factors]
+
+        """
+        session_maxlen = sessions.shape[1]
+        attn_mask = None
+        if self.use_causal_attn:
+            attn_mask = ~torch.tril(
+                torch.ones((session_maxlen, session_maxlen), dtype=torch.bool, device=sessions.device)
+            )
+        timeline_mask = (sessions != 0).unsqueeze(-1)  # [batch_size, session_maxlen, 1]
+        seqs = item_embs[sessions]  # [batch_size, session_maxlen, factors]
+        seqs = self.pos_encoding(seqs, timeline_mask)
+        seqs = self.emb_dropout(seqs)
+        seqs = self.transformer_blocks(seqs, timeline_mask, attn_mask)
+        return seqs
+
+    def forward(
+        self,
+        sessions: torch.Tensor,  # [batch_size, session_maxlen]
+    ) -> torch.Tensor:
+        """TODO"""
+        item_embs = self.item_model.get_all_embeddings()  # [n_items + 1, factors]
+        session_embs = self.encode_sessions(sessions, item_embs)  # [batch_size, session_maxlen, factors]
+        logits = session_embs @ item_embs.T  # [batch_size, session_maxlen, n_items + 1]
+        return logits
+
+
+# ####  --------------  Trainer  --------------  #### #
+
+
+class Trainer:
+    """TODO"""
+
+    def __init__(
+        self,
+        lr: float,
+        epochs: int,
+        device: torch.device,
+        loss: str = "softmax",
+    ):
+        """TODO"""
+        self.model: TransformerBasedSessionEncoder
+        self.optimizer: torch.optim.Adam
+        self.lr = lr
+        self.epochs = epochs
+        self.device = device
+        self.loss_func = self._init_loss_func(loss)  # TODO: move loss func to `SasRec` class
+
+    def fit(
+        self,
+        model: TransformerBasedSessionEncoder,
+        fit_dataloader: DataLoader,
+    ) -> None:
+        """TODO"""
+        self.model = model
+        self.optimizer = self._init_optimizers()
+        self.model.to(self.device)
+
+        self.xavier_normal_init(self.model)
+        self.model.train()  # enable model training
+
+        # self.model.item_model.to_device(self.device)
+
+        epoch_start_idx = 1
+
+        # ce_criterion = torch.nn.CrossEntropyLoss()
+        # https://github.com/NVIDIA/pix2pixHD/issues/9 how could an old bug appear again...
+
+        try:
+            for epoch in range(epoch_start_idx, self.epochs + 1):
+                logger.info("training epoch %s", epoch)
+                for x, y, w in fit_dataloader:
+                    x = x.to(self.device)  # [batch_size, session_maxlen]
+                    y = y.to(self.device)  # [batch_size, session_maxlen]
+                    w = w.to(self.device)  # [batch_size, session_maxlen]
+
+                    self.train_step(x, y, w)
+
+        except KeyboardInterrupt:
```

Comment:
We don't really need this try..except block

Reply:
Agreed. Trainer is being completely refactored right now. We didn't do it yet. And we are also moving to Lightning module and trainer in one of the nearest PRs.

Reply:
ok, just add TODO please

Reply:
removed it